src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,860 @@
+            for (sub_datum_ref, sub_field) in struct_ref.iter_fields_ref().zip_eq_debug(
+                st.iter()
+                    .map(|(name, dt)| Field::with_name(dt.clone(), name)),
+            ) {
+                doc.insert(
+                    sub_field.name.clone(),
+                    bson_from_scalar_ref(&sub_field, sub_datum_ref),
+                );
+            }
+            Bson::Document(doc)
+        }
+        (DataType::List(dt), ScalarRefImpl::List(v)) => {
+            let inner_field = Field::unnamed(Box::<DataType>::into_inner(dt));
+            v.iter()
+                .map(|scalar_ref| bson_from_scalar_ref(&inner_field, scalar_ref))
+                .collect::<Bson>()
+        }
+        (DataType::Bytea, ScalarRefImpl::Bytea(v)) => Bson::Binary(Binary {
+            subtype: BinarySubtype::Generic,
+            bytes: v.into(),
+        }),
+        _ => {
+            if let Ok(suppressed_count) = LOG_SUPPERSSER.check() {
+                tracing::warn!(
+                    suppressed_count,
+                    ?field,
+                    ?scalar_ref,
+                    "bson_from_scalar_ref: unsupported data type"
+                );
+            }
+            Bson::Null
+        }
+    }
+}
+
+struct InsertCommandBuilder {
+    coll: String,
+    inserts: Array,
+}
+
+impl InsertCommandBuilder {
+    fn new(coll: String, capacity: usize) -> Self {
+        Self {
+            coll,
+            inserts: Array::with_capacity(capacity),
+        }
+    }
+
+    fn append(&mut self, row: Document) {
+        self.inserts.push(Bson::Document(row));
+    }
+
+    fn build(self) -> Document {
+        doc! {
+            "insert": self.coll,
+            "ordered": true,
+            "documents": self.inserts,
+        }
+    }
+}
+
+struct UpsertCommandBuilder {
+    coll: String,
+    upserts: Array,
+    deletes: HashMap<Vec<u8>, Document>,
+}
+
+impl UpsertCommandBuilder {
+    fn new(coll: String, capacity: usize) -> Self {
+        Self {
+            coll,
+            upserts: Array::with_capacity(capacity),
+            deletes: HashMap::with_capacity(capacity),
+        }
+    }
+
+    fn add_upsert(&mut self, pk: Document, row: Document) -> Result<()> {
+        let pk_data = mongodb::bson::to_vec(&pk)
+            .map_err(|err| anyhow!(err).context("cannot serialize primary key"))?;
+        // under same pk, if the record currently being upserted was marked for deletion previously, we should
+        // revert the deletion, otherwise, the upserting record may be accidentally deleted.
+        // see https://github.com/risingwavelabs/risingwave/pull/17102#discussion_r1630684160 for more information.
+        self.deletes.remove(&pk_data);
+
+        self.upserts.push(bson!( {
+            "q": pk,
+            "u": row,
+            "upsert": true,
+            "multi": false,
+        }));
+
+        Ok(())
+    }
+
+    fn add_delete(&mut self, pk: Document) -> Result<()> {
```

Comment:
We may need to also remove the upsert if there is any.

Reply:
I think reverting the delete during upsert is enough, as we have ignored the `UpdateDelete` in upsert mode. Thus there should not be an inconsistency problem with `insert pk delete pk` as mentioned by @xxhZs. And also this is a temporary solution due to the absence of `BulkWrite` API in the mongodb driver. We may need to rewrite all `CommandBuilder` related code when mongodb driver supports `BulkWrite`, in order to simplify our code in the future.